feat(orchestrator): enable custom auth provider

batzionb opened the pull request:
resolves [FLPATH-2416](https://issues.redhat.com/browse/FLPATH-2416)

gciavarrini commented on workspaces/orchestrator/docs/auth.md:
```diff
@@ -6,21 +6,31 @@ This guide explains how to declare authentication requirements in a SonataFlow w
+| gitHub     | `X-Authorization-Github`     | Built-in |
+| gitLab     | `X-Authorization-Gitlab`     | Built-in |
+| microsoft  | `X-Authorization-Microsoft`  | Built-in |
+| github-two | `X-Authorization-Github-Two` | Custom   |
```
WDYT about renaming `github-two` to `github-custom` for clarity?

batzionb replied:
This would change the custom plugins introduced in this PR https://github.com/redhat-developer/rhdh-plugins/pull/1138 related to auth for widgets library and docs of the widgets library, I don't want to include changes in those libraries in the scope of this PR. Perhaps we can take this with @mareklibra and resolve in a different PR

gciavarrini commented on workspaces/orchestrator/plugins/orchestrator/src/api/OrchestratorAuth.ts:
```diff
@@ -13,86 +13,151 @@
+      // Find the API reference for the custom provider
+      const apiRef = allPlugins
+        .flatMap(plugin => Array.from(plugin.getApis()))
+        .filter((api: AnyApiFactory) => api.api.id === providerApiId)
+        .at(0)?.api as ApiRef<OpenIdConnectApi & OAuthApi> | undefined;
```
you can simplify this with 
```suggestion
        .find((api: AnyApiFactory) => api.api.id === providerApiId)
        ?.api as ApiRef<OpenIdConnectApi & OAuthApi> | undefined;
```

gciavarrini commented on workspaces/orchestrator/plugins/orchestrator/src/api/OrchestratorAuth.ts:
```diff
@@ -13,86 +13,151 @@
+          authApi = microsoftAuthApi;
+          break;
+        default:
+          if (!tokenDescriptor.custonmProviderApiId) {
```
```suggestion
          if (!tokenDescriptor.customProviderApiId) {
```

batzionb replied:
fixed

gciavarrini commented on workspaces/orchestrator/docs/auth.md:
```diff
@@ -41,17 +51,18 @@ quarkus.openapi-generator.github_yaml.auth.BearerToken.header-name=X-Authorizati
+| `ui:widget`                     | string             | Yes                  | Must be `"AuthRequester"`                                                                                                                                                                        |
+| `ui:props.authTokenDescriptors` | array of objects   | Yes                  | List of token requirements                                                                                                                                                                       |
+| — `provider`                    | string             | Yes                  | Built-in: `github`, `gitlab`, `microsoft` or custom provider identifier                                                                                                                          |
+| — `custonmProviderApiId`        | string             | For custom providers | Backstage ApiRef id of the custom provider plugin (e.g., `my.custom.auth.github-two` from the [custom-authentication-provider-module](../plugins/custom-authentication-provider-module/) plugin) |
```
typo `s/custonmProviderApiId/customProviderApiId/`

batzionb replied:
fixed

gciavarrini commented on workspaces/orchestrator/docs/auth.md:
```diff
@@ -66,14 +77,22 @@ quarkus.openapi-generator.github_yaml.auth.BearerToken.header-name=X-Authorizati
+        },
+        {
+          "provider": "github-two",
+          "custonmProviderApiId": "my.custom.auth.github-two",
```
typo?

gciavarrini commented on workspaces/orchestrator/plugins/orchestrator/src/api/OrchestratorAuth.ts:
```diff
@@ -13,86 +13,151 @@
+      }
+
+      const authTokens: Array<AuthToken> = [];
+      for (const tokenDescriptor of tokenDescriptors) {
+        authTokens.push(await getToken(tokenDescriptor));
+      }
```
Do we need to sequentially check them or is it possible to use `Promise.all`?

batzionb replied:
It must be sequential because each getToken call potentially shows a pop up to the user, there can be multiple popups that the user needs to log in to

gciavarrini reviewed (commented):
some comments.
In addition, can you fix the Sonar "2 new issues"?

batzionb commented:
Wow, thanks for finding the typo!

batzionb commented:
@gciavarrini In addition to the sonar fixes and the fixes following your review (thanks!) I included some more small fixes:

1. Made tokenType optional since oauth is more common
2. Added typescript typeguard to encapsulate all the type checks
3. Moved the hook file to the correct directory
4. Some doc improvements

Thanks again!

gciavarrini reviewed (approved):
lgtm
